### DISCUSSION TITLE
How to calculate the off-diagonal terms of the Jacobian in CoupledVarNeumannBC

### DISCUSSION THREAD

**Discussion by aaelmeli** [idaholab/moose#19465] [Q&A General] [answered]
Hi

I recently encountered a problem where I have two Linear Systems that are coupled only on the boundaries.  The `CoupledVarNeumannBC`  object was used to achieve the coupling. The issue is, `CoupledVarNeumannBC` does not have a method that provides the Jacobian. This means that, whenever a direct linear solver is used, it will not show any coupling because the contribution of the `CoupledVarNeumannBC`  to the off-diagonal terms in the Jacobian is zero. This enforces using a nonlinear solver where the Jacobian can be approximated numerically using for example FD. This issue was under discussion in #19365 where @GiudGiud had fruitful thoughts. The issue is, for some coupled PDEs, like the one in #19365, this would result in slow convergence or even divergence.

So, I am re-posting the following question here to have better visibility in case anyone encountered a similar problem.

The question is, Is it possible to implement a function that calculates the Jacobian terms of the `CoupledVarNeumannBC` object? Any thoughts on the implementation?

**Comment by lindsayad** [comment, marked as answer]
This class was added in https://github.com/idaholab/moose/pull/13503 for use with aux variables. We should update this class in order to provide Jacobian calculations if the coupled variable is a "nonlinear" variable.

**Reply by lindsayad** [reply]
I've created https://github.com/idaholab/moose/issues/19468 to address this

**Reply by aaelmeli** [reply]
> I've created #19468 to address this

Thank you.

**Reply by GiudGiud** [reply]
@aaelmeli Alex made a PR and it is merged in the framework. Can you please try again now?

**Reply by aaelmeli** [reply]
> @aaelmeli Alex made a PR and it is merged in the framework. Can you please try again now?

I have tested it, it works just fine. Thank you, @lindsayad and @GiudGiud
